alf/algorithms/rl_algorithm.py

```diff
@@ -478,6 +479,74 @@ def _rollout_step(self, time_step: TimeStep, state):
     @common.mark_rollout
     @data_distributed_when(lambda algorithm: algorithm.on_policy)
     def unroll(self, unroll_length: int):
+        if not self._config.async_unroll:
+            return self._unroll(unroll_length)
+        if self._async_unroller is None:
+            # self._env.close()
+            self._async_unroller = AsyncUnroller(
+                self, self._config.unroll_queue_size, self._config.root_dir,
+                common.get_conf_file())
+        elif alf.summary.get_global_counter(
+        ) % self._config.unroll_parameter_update_period == 0:
+            self._async_unroller.update_parameter(self)
+
+        assert not self._overwrite_policy_output, (
+            "async_unroll does not "
+            "support overwrite_policy_output")
+        assert not self.on_policy, ("async_unroll does not support on-policy "
+                                    "training")
+        experience_list = []
+        original_reward_list = []
+        env_step_time = 0.
+        store_exp_time = 0.
+        unroll_results = self._async_unroller.gather_unroll_results(
+            unroll_length, self._config.max_unroll_length)
+        if self._rollout_info_spec is None and len(unroll_results) > 0:
+            self._rollout_info_spec = dist_utils.extract_spec(
+                unroll_results[0].policy_step.info)
+
+        for unroll_result in unroll_results:
+            time_step = unroll_result.time_step
+            policy_step = unroll_result.policy_step
+            policy_state = unroll_result.policy_state
+
+            env_step_time += unroll_result.env_step_time
+            self.observe_for_metrics(time_step.cpu())
+            exp = make_experience(time_step.cpu(), policy_step, policy_state)
+
+            t0 = time.time()
+            self.observe_for_replay(exp)
+            store_exp_time += time.time() - t0
+
+            exp_for_training = Experience(
+                time_step=unroll_result.transformed_time_step,
+                action=unroll_result.policy_step.output,
+                rollout_info=dist_utils.distributions_to_params(
+                    policy_step.info),
+                state=policy_state)
+
+            experience_list.append(exp_for_training)
+            original_reward_list.append(time_step.reward)
```

Comment:
Feel like this part of making experiences can be merged with the corresponding part in ``self._unroll`` to reduce code maintenance?

Reply:
That's true. But I also feel making more abstraction will lower the readability. There are some code specifically for async_unroll after the loop.